ext/CatalystHomotopyContinuationExtension/homotopy_continuation_extension.jl

```diff
@@ -0,0 +1,124 @@
+
+Examples:
+```@repl
+rs = @reaction_network begin
+    k1, Y --> 2X
+    k2, 2X --> X + Y
+    k3, X + Y --> Y
+    k4, X --> 0
+end
+ps = [:k3 => 1.0, :k2 => 2.0, :k4 => 1.5, :k1=>8.0]
+hc_sol = hc_steady_states(rs, ps)
+```
+gives
+```
+[0.5000000000000002, 2.0000000000000004]
+[0.0, 0.0]
+[4.499999999999999, 5.999999999999999]
+```
+
+Notes:
+- This is a wrapper around the `solve` function provided by HomotopyContinuation.jl, all credit for this functionality to that package's authors.
+```
+  """
+function Catalyst.hc_steady_states(rs::ReactionSystem, ps; filter_negative=true, neg_thres=-1e-20, u0=typeof(ps)(), kwargs...)
+    ss_poly = steady_state_polynomial(rs, ps, u0)
+    sols = HC.real_solutions(HC.solve(ss_poly; kwargs...))
+    reorder_sols!(sols, ss_poly, rs)
+    return (filter_negative ? filter_negative_f(sols; neg_thres=neg_thres) : sols)
+end
+
+# For a given reaction system, paraemter values, and initial conditions, find the polynomial that HC solves to find steady states.
+function steady_state_polynomial(rs::ReactionSystem, ps, u0)
+    ns = convert(NonlinearSystem, rs; remove_conserved = true)
+    pre_varmap = map(pair -> pair::Pair{Num,Float64}, [symmap_to_varmap(rs,u0); symmap_to_varmap(rs,ps)]) 
+    conservationlaw_errorcheck(rs, pre_varmap)
+    p_vals = ModelingToolkit.varmap_to_vars(pre_varmap, parameters(ns); defaults = ModelingToolkit.defaults(ns))
+    p_dict  = Dict(parameters(ns) .=> p_vals)
+    eqs_pars_funcs = vcat(equations(ns), conservedequations(rs))
+    eqs_funcs = map(eq -> substitute(eq.rhs - eq.lhs, p_dict), eqs_pars_funcs)
+    eqs = [deregister([mm, mmr, hill, hillr, hillar], eq) for eq in eqs_funcs]
+    eqs_intexp = make_int_exps.(eqs)
+    return Catalyst.to_multivariate_poly(remove_denominators.(eqs_intexp))
+end
+
+# If u0s are not given while conservation laws are present, throws an error.
+function conservationlaw_errorcheck(rs, pre_varmap)
+    vars_with_vals = union(first.(pre_varmap), keys(ModelingToolkit.defaults(rs)))
+    isempty(intersect(species(rs), vars_with_vals)) || return
+    isempty(conservedequations(rs)) && return 
+    error("The system have conservation laws but no initial conditions were provided. Please provide initial conditions.")
+end
+
+# Unfolds a function (like mm or hill). 
+function deregister(fs::Vector{T}, expr) where T
+    for f in fs
+        expr = deregister(f, expr) 
+    end
+    return expr
+end
+# Provided by Shashi Gowda.
+deregister(f, expr) = wrap(Rewriters.Postwalk(Rewriters.PassThrough(___deregister(f)))(unwrap(expr)))
+function ___deregister(f)
+    (expr) ->
+    if istree(expr) && operation(expr) == f
+        args = arguments(expr)
+        invoke_with = map(args) do a
+            t = typeof(a)
+            issym(a) || istree(a) ? wrap(a) => symtype(a) : a => typeof(a)
+        end 
+        invoke(f, Tuple{last.(invoke_with)...}, first.(invoke_with)...)
+    end
+end
```

Comment:
Calling invoke seems hacky. 

Perhaps we should just not register the functions to begin with? Was there any good reason we did so originally?

Reply:
I guess such a change might need to wait for the next breaking release though.

Reply:
I and Shashi sat together and fixed this (well Shashi mostly). He was interested in improving this though, so we opened an issue to hopefully have some native support fro this in Symbolics: https://github.com/JuliaSymbolics/Symbolics.jl/issues/975